File: packages/snow_draw_core/lib/draw/edit/core/edit_modifiers.dart
import 'package:meta/meta.dart';

/// Edit-time keyboard modifiers coming from the UI/input layer.
///
/// These flags represent transient modifier keys at the time an edit update is
/// processed.
@immutable
class EditModifiers {
  const EditModifiers({
    this.maintainAspectRatio = false,
    this.fromCenter = false,
    this.discreteAngle = false,
    this.snapOverride = false,
  });

  /// Keeps the edited bounds constrained to an aspect ratio when supported.
  final bool maintainAspectRatio;

  /// Resizes or transforms relative to the center when supported.
  final bool fromCenter;

  /// Applies discrete rotation snapping when supported.
  final bool discreteAngle;

  /// Overrides default snapping behavior when supported.
  final bool snapOverride;
}

/// Policy for handling update failures during edit sessions.
enum EditUpdateFailurePolicy { toIdle, keepState }

File: packages/snow_draw_rust_engine/src/draw/edit/core/edit_modifiers.rs
/// Edit-time keyboard modifiers coming from the UI/input layer.
///
/// These flags represent transient modifier keys at the time an edit update is
/// processed.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash)]
pub struct EditModifiers {
    /// Keeps the edited bounds constrained to an aspect ratio when supported.
    pub maintain_aspect_ratio: bool,

    /// Resizes or transforms relative to the center when supported.
    pub from_center: bool,

    /// Applies discrete rotation snapping when supported.
    pub discrete_angle: bool,

    /// Overrides default snapping behavior when supported.
    pub snap_override: bool,
}

impl EditModifiers {
    /// Creates a new set of edit modifiers.
    pub const fn new(
        maintain_aspect_ratio: bool,
        from_center: bool,
        discrete_angle: bool,
        snap_override: bool,
    ) -> Self {
        Self {
            maintain_aspect_ratio,
            from_center,
            discrete_angle,
            snap_override,
        }
    }
}
